### PR TITLE
Compute fastANI by column (many queries vs one reference)

### LINKED ISSUE #180: How to run large scale fastANI comparisons quicker?
## Summary:

Initially we are running a separate fastANI command for each pairwise comparison, meaning for 1000 genomes we would make a million command line calls and associated individual additions to the DB. This does not scale well.

The tool itself can take a single query, or a list of queries, and a single reference (subject) or a list of references (subjects). This allows us to consider various batching strategies taking into account run time and memory usage.

Using many queries against a single reference (subject) works best, with runtime scaling linearly, and RAM usage as per a single comparison.

## Measurements

Sample timings on a 2015 Intel Apple iMac with 8GB of physical RAM, and 8 cores. All these examples are run single threaded. Memory usage as reported by the macOS Activity Monitory in the GUI. The genomes are bacteria.

```console
$ fastANI -v
version 1.33
```

### One query vs One reference:

Timing a single comparison, typically about 1s, memory usage modest:

```console
$ time fastANI -q pecto_md5/bfbb456d1f9919a3ca833506babecdae.fasta -r pecto_md5/c56f7321b4dc97878a0af3941ae4c4e0.fasta  -o /tmp/one_vs_one.tsv 2>/dev/null ; echo "Return code $?"
fastANI -q pecto_md5/bfbb456d1f9919a3ca833506babecdae.fasta -r  -o  2>   1.05s user 0.02s system 97% cpu 1.096 total
Return code 0
$ cat /tmp/one_vs_one.tsv
pecto_md5/bfbb456d1f9919a3ca833506babecdae.fasta        pecto_md5/c56f7321b4dc97878a0af3941ae4c4e0.fasta        92.2189 1257    1581
```

### Many queries vs One reference:

25 queries (117M total) vs one reference, observed peak memory approx 45M, about 20s:

```console
$ export N=25; time fastANI --ql pecto_md5_$N.txt -r pecto_md5/bfbb456d1f9919a3ca833506babecdae.fasta -o /tmp/$N_vs_one.tsv 2>/dev/null ; echo "Return code $?" ; wc -l /tmp/$N_vs_one.tsv
fastANI --ql pecto_md5_$N.txt -r  -o /tmp/$N_vs_one.tsv 2> /dev/null  19.91s user 0.23s system 99% cpu 20.228 total
Return code 0
      25 /tmp/.tsv
```

100 queries (476M total) vs one reference, observed peak memory approx 43M, about 1m20s or ~80s:

```console
$ export N=100; time fastANI --ql pecto_md5_$N.txt -r pecto_md5/bfbb456d1f9919a3ca833506babecdae.fasta -o /tmp/$N_vs_one.tsv 2>/dev/null ; echo "Return code $?" ; wc -l /tmp/$N_vs_one.tsv
fastANI --ql pecto_md5_$N.txt -r  -o /tmp/$N_vs_one.tsv 2> /dev/null  76.81s user 0.77s system 97% cpu 1:19.27 total
Return code 0
     100 /tmp/.tsv
```

200 queries (944M total) vs one reference, observed peak memory approx 41M, about 2m45s or ~165s:

```console
$ export N=200; time fastANI --ql pecto_md5_$N.txt -r pecto_md5/bfbb456d1f9919a3ca833506babecdae.fasta -o /tmp/$N_vs_one.tsv 2>/dev/null ; echo "Return code $?" ; wc -l /tmp/$N_vs_one.tsv
fastANI --ql pecto_md5_$N.txt -r  -o /tmp/$N_vs_one.tsv 2> /dev/null  155.02s user 1.85s system 94% cpu 2:46.26 total
Return code 0
     200 /tmp/.tsv
```

300 queries (944M total) vs one reference, observed peak memory approx 44M, almost 4m or 240s:

```console
$ export N=300; time fastANI --ql pecto_md5_$N.txt -r pecto_md5/bfbb456d1f9919a3ca833506babecdae.fasta -o /tmp/$N_vs_one.tsv 2>/dev/null ; echo "Return code $?" ; wc -l /tmp/$N_vs_one.tsv
fastANI --ql pecto_md5_$N.txt -r  -o /tmp/$N_vs_one.tsv 2> /dev/null  230.07s user 2.51s system 99% cpu 3:53.40 total
Return code 0
     300 /tmp/.tsv
```

So using many queries with a single references scales fine for CPU run time (about 0.8s per comparison), and the memory usage is stable at about 40MB (with bacteria).

### One query vs Many references:

One query vs 25 refs (117M total), observed peak memory approx 392M, about 23s:

```console
export N=25; time fastANI -q pecto_md5/bfbb456d1f9919a3ca833506babecdae.fasta --rl pecto_md5_$N.txt -o /tmp/one_vs_$N.tsv 2>/dev/null ; echo "Return code $?" ; wc -l /tmp/one_vs_$N.tsv
fastANI -q pecto_md5/bfbb456d1f9919a3ca833506babecdae.fasta --rl  -o  2>   21.71s user 0.61s system 97% cpu 22.789 total
Return code 0
      25 /tmp/one_vs_25.tsv
```

One query vs 100 refs (476M total), observed peak memory approx 1.25GB, about 1m30s or ~90s:

```console
$ export N=100; time fastANI -q pecto_md5/bfbb456d1f9919a3ca833506babecdae.fasta --rl pecto_md5_$N.txt -o /tmp/one_vs_$N.tsv 2>/dev/null ; echo "Return code $?" ; wc -l /tmp/one_vs_$N.tsv
fastANI -q pecto_md5/bfbb456d1f9919a3ca833506babecdae.fasta --rl  -o  2>   79.79s user 1.51s system 90% cpu 1:29.84 total
Return code 0
     100 /tmp/one_vs_100.tsv
```

One query vs 200 refs (944M total), observed peak memory approx 2.28G, about 3 mins 10s, or ~190s:

```console
$ export N=200; time fastANI -q pecto_md5/bfbb456d1f9919a3ca833506babecdae.fasta --rl pecto_md5_$N.txt -o /tmp/one_vs_$N.tsv 2>/dev/null ; echo "Return code $?" ; wc -l /tmp/one_vs_$N.tsv
fastANI -q pecto_md5/bfbb456d1f9919a3ca833506babecdae.fasta --rl  -o  2>   160.17s user 3.71s system 86% cpu 3:09.93 total
Return code 0
     200 /tmp/one_vs_200.tsv
```

One query vs 300 refs (1.4G total), observed peak memory approx 3.21G, about 4mins30s or ~270s:

```console
$ export N=300; time fastANI -q pecto_md5/bfbb456d1f9919a3ca833506babecdae.fasta --rl pecto_md5_$N.txt -o /tmp/one_vs_$N.tsv 2>/dev/null ; echo "Return code $?" ; wc -l /tmp/one_vs_$N.tsv
fastANI -q pecto_md5/bfbb456d1f9919a3ca833506babecdae.fasta --rl  -o  2>   236.41s user 5.93s system 88% cpu 4:33.79 total
Return code 0
     300 /tmp/one_vs_300.tsv
```

So using a single query and many references scales fine for CPU run time (approx 0.9s per comparison - slightly slower), but is dreadful for memory usage (over double the reference size on disk?).

This means one-vs-many is impractical on this machine due to RAM limits, but moreover the memory usage becomes problematic for doing this kind of analysis on an HPC cluster.

This has been noted on https://github.com/ParBLiSS/FastANI/issues/76 and batching the tasks is the suggested remedy. i.e. I don't expect fastANI to cater to this usage pattern.

### Many queries vs Many references:

Based on the above we can estimate the run time as 0.9s per pair, and expect the memory usage to scale as for the one query vs many references case. Thus 100 genomes, 10000 pairs, under 3 hours and 1.25GB peak memory?

Or more manageable, 25 genomes (117M on disk) should take well under 10mins? Peak memory observed 393MB, and about 7 mins:

```console
$ export N=25; time fastANI --ql pecto_md5_$N.txt --rl pecto_md5_$N.txt -o /tmp/$N_vs_$N.tsv 2>/dev/null ; echo "Return code $?" ; wc -l /tmp/$N_vs_$N.tsv
fastANI --ql pecto_md5_$N.txt --rl pecto_md5_$N.txt -o /tmp/$N_vs_$N.tsv 2>   415.53s user 3.40s system 99% cpu 7:02.97 total
Return code 0
     625 /tmp/25.tsv
```

## Conclusion

As expected, run time wise, fastANI 1.33 scales linearly with the number of queries and/or references. However, using multiple references (``fastANI --rl ...``) is not recommended due to excessive RAM usage. If used, this should be tightly limited to a small number based on the available RAM.

Note that fastANI's built in threading support _requires_ the use of multiple references, see https://github.com/ParBLiSS/FastANI/pull/140

While we could split a large comparison into smaller blocks (e.g. 25 x 25 genomes ought to take about 10 mins on this machine), doing N queries vs 1 reference is simpler and does not need excessive RAM.

Ignoring overheads including DB logging, this means a 1000 genome comparison should take under a million seconds, or under 11 days CPU time - and is easily split into 1000 separate jobs (one per reference, each expected to take about 15 mins). Depending on the cluster size and availability, you could expect this to be done in a few hours.

Note in any case, note that the output file is not written incrementally, but only at the very end of the job. This is frustrating as we could parse this as written, capturing any partial results if the job is interrupted, and also providing progress feedback to the user (i.e. updating our progress bar). i.e. With this example we'd not expect any results back until at least 15 minutes pass. However, we can then record those 1000 results as a single bulk import to our DB which will be much faster, reducing our overhead considerably.

### PR DESCRIPTION
**Pull request by peterjc**
Closes #180.

## Type of change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality not to work as expected)
- [ ] This change requires a documentation update
- [ ] This is a documentation update

## Action Checklist

- [x] Work on a single issue/concept (if there are multiple separate issues to address, please use a separate pull request for each)
- [ ] Fork the `pyani-plus` repository under your own account (please [allow write access for repository maintainers](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork))
- [x] Set up an appropriate development environment (please see `CONTRIBUTING.md` for this repository)
- [x] Create a new branch with a short, descriptive name
- [x] Work on this branch
  - [x] style guidelines have been followed (please see `CONTRIBUTING.md` for this repository)
  - [x] new code is commented, especially in hard-to-understand areas
  - [x] corresponding changes to documentation have been made
  - [x] tests for the change have been added that demonstrate the fix or feature works
- [x] Test locally with `pytest-plus -v` **non-passing code will not be merged**
- [x] Rebase against `origin/master`
- [x] Check changes with linting/formatting tools before submission (please see `CONTRIBUTING.md` for this repository)
- [x] Commit branch
- [x] Submit pull request, describing the content and intent of the pull request
- [x] Request a code review
- [x] Continue the discussion at [`Pull requests` section](https://github.com/widdowquinn/pyani-plus/pulls) in the `pyan-plus` repository

### CODE REVIEW

File: pyani_plus/private_cli.py [hunk 1]
```diff
@@ -427,64 +425,180 @@ def log_fastani(  # noqa: PLR0913
...
+    for md5, filename in hash_to_filename.items():
+        if md5 not in completed_queries:
+            print(fasta_dir / filename)
+            empty = False
+    if empty:
+        if include_subject:
+            print(fasta_dir / hash_to_filename[subject_hash])
+        if not quiet:
+            sys.stderr.write(
+                f"INFO: Already have all comparisons for subject {subject_hash}\n"
```
Comment (peterjc) [inline pyani_plus/private_cli.py:600]: Need a test for this corner case

File: pyani_plus/public_cli.py [hunk 2]
```diff
@@ -588,7 +589,13 @@ def resume(  # noqa: C901, PLR0912, PLR0915
             binaries = {
                 "fastani": tool.exe_path,
             }
-            target_extension = ".fastani"
+            # Ideally this would be selective, as this will recompute every column
+            # (but we minimise this to a self-versus-self as fastANI won't no-op):
+            targets = [
+                f"all_vs_{Path(_.fasta_filename).stem}.fastani"
+                for _ in run.fasta_hashes
+            ]
```
Comment (peterjc) [inline pyani_plus/public_cli.py:597]: We could examine the full set of comparisons here to determine which subjects are already complete. i.e. which columns of the database do not need to be computed.

This would avoid the problematic behaviour of fastANI not liking being asked to do nothing vs a reference genome.

**Comment by peterjc** [comment]
Will want to rebase this after #186 which solves pre-existing comparisons in the DB more elegantly and I assume more efficiently.

**Review by widdowquinn — APPROVED** [review]
Very happy with the direction - if you need to polish we can look again.
